plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -178,36 +178,61 @@ type CekValEnv uni fun = UniqueMap TermUnique (CekValue uni fun)
 -- defers to the function stored in the environment). This makes the budgeting machinery extensible
 -- and allows us to separate budgeting logic from evaluation logic and avoid branching on the union
 -- of all possible budgeting state types during evaluation.
-newtype CekBudgetSpender cost uni fun s = CekBudgetSpender
-    { unCekBudgetSpender
-        :: ExBudgetCategory fun -> ExBudget -> CekM cost uni fun s ()
+newtype CekBudgetSpender fun s = CekBudgetSpender
+    { unCekBudgetSpender :: ExBudgetCategory fun -> ExBudget -> CekM s ()
     }
 
 -- General enough to be able to handle a spender having one, two or any number of 'STRef's
 -- under the hood.
 -- | Runtime budgeting info.
-data ExBudgetInfo cost uni fun s = ExBudgetInfo
-    { _exBudgetModeSpender  :: !(CekBudgetSpender cost uni fun s)  -- ^ A spending function.
+data ExBudgetInfo cost fun s = ExBudgetInfo
+    { _exBudgetModeSpender  :: !(CekBudgetSpender fun s)  -- ^ A spending function.
     , _exBudgetModeGetFinal :: !(ST s cost)                        -- ^ For accessing the final state.
     }
 
 -- We make a separate data type here just to save the caller of the CEK machine from those pesky
 -- 'ST'-related details.
+-- Strictly speaking the 'uni' param is unneeded, but it helps avoid places where we'd otherwise need
+-- a 'Proxy' to pin down 'uni'.
 -- | A budgeting mode to execute the CEK machine in.
-newtype ExBudgetMode cost uni fun = ExBudgetMode
-    { unExBudgetMode :: forall s. ST s (ExBudgetInfo cost uni fun s)
+newtype ExBudgetMode cost (uni :: Type -> Type) fun = ExBudgetMode
```

Comment:
Nitpicking: I'd import `Data.Kind` qualified as `GHC` and make it `uni :: GHC.Type -> GHC.Type` like we do in other Plutus-specific places. But that hardly matters.

Reply:
Good idea